torch/cuda/amp/autocast_mode.py

```diff
@@ -112,31 +122,37 @@ def forward(self, input):
     Args:
         enabled(bool, optional, default=True):  Whether autocasting should be enabled in the region.
     """
-    def __init__(self, enabled=True):
-        if enabled and not torch.cuda.is_available():
+
+    def __init__(self, enabled: bool = True):
+        if torch._jit_internal.is_scripting():
+            self._enabled = enabled
+        elif enabled and not torch.cuda.is_available():
             warnings.warn("torch.cuda.amp.autocast only affects CUDA ops, but CUDA is not available.  Disabling.")
             self._enabled = False
         else:
             self._enabled = enabled
 
     def __enter__(self):
+        if torch._jit_internal.is_scripting():
+            return self
         self.prev = torch.is_autocast_enabled()
         torch.set_autocast_enabled(self._enabled)
         torch.autocast_increment_nesting()
 
-    def __exit__(self, *args):
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any):
+        if torch._jit_internal.is_scripting():
+            return
         # Drop the cache when we exit to a nesting level that's outside any instance of autocast.
         if torch.autocast_decrement_nesting() == 0:
             torch.clear_autocast_cache()
         torch.set_autocast_enabled(self.prev)
         return False
 
     def __call__(self, func):
-        @functools.wraps(func)
-        def decorate_autocast(*args, **kwargs):
-            with self:
-                return func(*args, **kwargs)
-        return decorate_autocast
+        if torch._jit_internal.is_scripting():
+            return func
```

Comment:
Should we raise an error here to tell the user autocast as decorator under scripting doesn't work?

Reply:
This is not going to be invoked, we just need to add a definition which compiles fine with the TorchScript frontend.

PS. We do report an error, but through a different mechanism:
https://github.com/csarofeen/pytorch/blob/692e2361f68c1f5a869d46c41e5b8f87bef3d353/torch/cuda/amp/autocast_mode.py#L17